baseapp/abci.go

```diff
@@ -1223,19 +1229,57 @@ func (app *BaseApp) CreateQueryContext(height int64, prove bool) (sdk.Context, e
 			)
 	}
 
-	// when a client did not provide a query height, manually inject the latest
-	if height == 0 {
-		height = lastBlockHeight
-	}
-
-	if height <= 1 && prove {
+	if height > 0 && height <= 1 && prove {
 		return sdk.Context{},
 			errorsmod.Wrap(
 				sdkerrors.ErrInvalidRequest,
 				"cannot query with proof when height <= 1; please provide a valid height",
 			)
 	}
 
+	var header *cmtproto.Header
+	for _, state := range []*state{
+		app.checkState,
+		app.finalizeBlockState,
+	} {
+		if state != nil {
+			// branch the commit multi-store for safety
+			h := state.Context().BlockHeader()
+			if height == 0 {
+				lastBlockHeight = qms.LatestVersion()
+			}
+			if !checkHeader || height != 0 || height == 0 && h.Height == lastBlockHeight {
+				header = &h
+				break
+			}
+		}
+	}
+
+	if header == nil {
+		return sdk.Context{},
+			errorsmod.Wrapf(
+				sdkerrors.ErrInvalidHeight,
+				"header height in all state context is not latest height (%d)", lastBlockHeight,
+			)
+	}
+
+	// when a client did not provide a query height, manually inject the latest
+	if height == 0 {
+		height = lastBlockHeight
+	} else {
+		defer func() {
+			if err == nil {
+				rms, ok := app.cms.(*rootmulti.Store)
+				if ok {
+					cInfo, err := rms.GetCommitInfo(height)
+					if cInfo != nil && err == nil {
+						ctx = ctx.WithHeaderInfo(coreheader.Info{Height: height, Time: cInfo.Timestamp})
+					}
+				}
+			}
+		}()
```

Comment:
why is this defer necessary?

Reply:
just replaced it with `isLatest` to avoid confusion